In [1]:
import pandas as pd
import numpy as np 
import sklearn


# Dataset: loading and initial inspection

In [2]:
mushroom=pd.read_csv('Mushroom.csv')
mushroom

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

In [4]:
mushroom.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

We notice that the column veil-type has only 1 unique value - that is, all 8124 mushroom instances have the same veil-color.

It thus becomes an irrelevant feature, so we proceed to remove it

In [6]:
mushroom.drop(['veil-type'], axis=1, inplace=True)

# Converting categorical data to numerical

In [7]:
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.preprocessing import LabelEncoder
lbl=LabelEncoder()
for col in mushroom.columns:
    mushroom[col] = lbl.fit_transform(mushroom[col])

In [8]:
mushroom

class  cap-shape  cap-surface  cap-color  bruises  odor  \
0         1          5            2          4        1     6   
1         0          5            2          9        1     0   
2         0          0            2          8        1     3   
3         1          5            3          8        1     6   
4         0          5            2          3        0     5   
...     ...        ...          ...        ...      ...   ...   
8119      0          3            2          4        0     5   
8120      0          5            2          4        0     5   
8121      0          2            2          4        0     5   
8122      1          3            3          4        0     8   
8123      0          5            2          4        0     5   

      gill-attachment  gill-spacing  gill-size  gill-color  ...  \
0                   1             0          1           4  ...   
1                   1             0          0           4  ...   
2                   1             0          0           5  ...   
3                   1             0          1           5  ...   
4                   1             1          0           4  ...   
...               ...           ...        ...         ...  ...   
8119                0             0          0          11  ...   
8120                0             0          0          11  ...   
8121                0             0          0           5  ...   
8122                1             0          1           0  ...   
8123                0             0          0          11  ...   

      stalk-surface-above-ring  stalk-surface-below-ring  \
0                            2                         2   
1                            2                         2   
2                            2                         2   
3                            2                         2   
4                            2                         2   
...                        ...                       ...   
8119                         2                         2   
8120                         2                         2   
8121                         2                         2   
8122                         2                         1   
8123                         2                         2   

      stalk-color-above-ring  stalk-color-below-ring  veil-color  ring-number  \
0                          7                       7           2            1   
1                          7                       7           2            1   
2                          7                       7           2            1   
3                          7                       7           2            1   
4                          7                       7           2            1   
...                      ...                     ...         ...          ...   
8119                       5                       5           1            1   
8120                       5                       5           0            1   
8121                       5                       5           1            1   
8122                       7                       7           2            1   
8123                       5                       5           1            1   

      ring-type  spore-print-color  population  habitat  
0             4                  2           3        5  
1             4                  3           2        1  
2             4                  3           2        3  
3             4                  2           3        5  
4             0                  3           0        1  
...         ...                ...         ...      ...  
8119          4                  0           1        2  
8120          4                  0           4        2  
8121          4                  0           1        2  
8122          0                  7           4        2  
8123          4                  4           1        2  

[8124 rows x 22 columns]

# Separating labels from features

In [10]:
y = mushroom['class'].to_frame()
X = mushroom.drop('class', axis=1)

In [11]:
y.head()

class
0      1
1      0
2      0
3      1
4      0

In [12]:
X.head()

cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0          5            2          4        1     6                1   
1          5            2          9        1     0                1   
2          0            2          8        1     3                1   
3          5            3          8        1     6                1   
4          5            2          3        0     5                1   

   gill-spacing  gill-size  gill-color  stalk-shape  ...  \
0             0          1           4            0  ...   
1             0          0           4            0  ...   
2             0          0           5            0  ...   
3             0          1           5            0  ...   
4             1          0           4            1  ...   

   stalk-surface-above-ring  stalk-surface-below-ring  stalk-color-above-ring  \
0                         2                         2                       7   
1                         2                         2                       7   
2                         2                         2                       7   
3                         2                         2                       7   
4                         2                         2                       7   

   stalk-color-below-ring  veil-color  ring-number  ring-type  \
0                       7           2            1          4   
1                       7           2            1          4   
2                       7           2            1          4   
3                       7           2            1          4   
4                       7           2            1          0   

   spore-print-color  population  habitat  
0                  2           3        5  
1                  3           2        1  
2                  3           2        3  
3                  2           3        5  
4                  3           0        1  

[5 rows x 21 columns]

# Standardising our features

It is generally considered a good practice to standardise our features (convert them to have zero-mean and unit variance). Most of the times, the difference will be small, but, in any case, it still never hurts to do so.

In [14]:
from sklearn.preprocessing import StandardScaler

In [15]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Creating training and sets sets

In [17]:
from sklearn.model_selection import train_test_split

In [18]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3,random_state=19)

# Logistic Regression

Since this is now a supervised learning binary classification problem, it makes perfect sense to start by running a simple logistic regression.

A logistic regression simply predicts the probability of an instance (row) belonging to the default class, which can then be snapped into a 0 or 1 classification. Off we go.


In [21]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [22]:
logreg = LogisticRegression()
logreg.fit(X_train, y_train.values.ravel())
y_pred_test = logreg.predict(X_test)
print('Accuracy of Logistic Regression classifier on the test set: {:.2f}'.format(accuracy_score(y_test, y_pred_test)))

Accuracy of Logistic Regression classifier on the test set: 0.96


we should run the logistic regression again, but this time using cross-validation, to ensure that we are not overfitting the data. A simple 10-fold cross validation should do.

In [26]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import StratifiedShuffleSplit

In [27]:
scores = cross_val_score(logreg, X_train, y_train.values.ravel(), cv=StratifiedShuffleSplit(n_splits=10, test_size=0.3, random_state=19), scoring='accuracy')
print('Accuracy of Logistic Regression classifier using 10-fold cross-validation: {}'.format(scores.mean()))


Accuracy of Logistic Regression classifier using 10-fold cross-validation: 0.9548065650644781


In [29]:
import joblib
joblib.dump(logreg,'mushroom.csv')

['mushroom.csv']

# Conclusion

We fitted a logistic regression model and achieved near perfect accuracy, so there was no need to try with more complex models.